### PR TITLE
ARROW-9362: [Java] Support reading/writing V5 MetadataVersion

### PR DESCRIPTION
**Pull request by lidavidm**
This also enables Flight to write differing metadata versions.

Not implemented: any checks for unions in read/write based on metadata version.

I refactored TestFileWriter very heavily as I wanted to ensure IpcOptions was thoroughly tested and didn't want to duplicate code. I hope this doesn't make the change too hard to review.

### CODE REVIEW

**Comment by pitrou** [comment]
Does Java already implement the required IPC union layout and semantics? Otherwise perhaps we should defer this PR until the union work is done.

**Comment by lidavidm** [comment]
We should wait for #7290, yes. (Is anyone reviewing it?)

Also, this now checks the metadata version against the schema before writing.

**Comment by lidavidm** [comment]
Would any Java maintainers be able to review? There's a bit of a circular dependency with this and #7290 - this is needed so that C++ will see the V5 metadata version and read with the right buffer layout, then #7290 will actually correct the buffer layout and enable the tests.

**Review by rymurr — APPROVED** [review]
LGTM. Awesome to get this in!

I like the DFS, I suppose the recursive algorithm wont cause a problem for any normal schema.

File: java/vector/src/test/java/org/apache/arrow/vector/ipc/MessageSerializerTest.java [hunk 1]
```diff
@@ -148,7 +150,7 @@ public void testSchemaDictionaryMessageSerialization() throws IOException {
   public ExpectedException expectedEx = ExpectedException.none();
 
   @Test
-  public void testSerializeRecordBatch() throws IOException {
+  public void testSerializeRecordBatchV4() throws IOException {
```
Comment (BryanCutler) [inline java/vector/src/test/java/org/apache/arrow/vector/ipc/MessageSerializerTest.java:153]: doesn't this need an `IpcOption` to set `MetadataVersion.V4`?

**Review by BryanCutler — COMMENTED** [review]
LGTM. I mostly looked at the vector ipc and tests, and just had a minor question.

File: java/vector/src/test/java/org/apache/arrow/vector/ipc/MessageSerializerTest.java (hunk 1 above)
Comment (lidavidm) [reply java/vector/src/test/java/org/apache/arrow/vector/ipc/MessageSerializerTest.java:153]: Good catch, thank you. I've added the IpcOption.

**Review by BryanCutler — APPROVED** [review]
LGTM

**Comment by BryanCutler** [comment]
So is this ok to merge before #7290 ?

**Comment by lidavidm** [comment]
> So is this ok to merge before #7290 ?

This should be OK. It doesn't enable unions in the integration tests and prevents reading/writing unions with the old version. (Of course, someone could build master and start writing files with V5 metadata and the wrong union layout?)

**Pull request closed without merging**